File: frontend/src/lib/types/runtime.ts
/** Runtime projection types — mirrors backend RuntimeProjectionService output. */

export interface RuntimeStep {
  step_id: string;
  status: string;
  action: string | null;
  tool_name: string | null;
  started_at: string | null;
  completed_at: string | null;
  error: string | null;
}

export interface RuntimeRun {
  run_id: string;
  plan_id: string | null;
  status: string;
  created_at: string | null;
  total_steps: number;
  completed_steps: number;
  progress_pct: number;
  blocking_step_id: string | null;
  blocking_reason: string | null;
}

export interface RuntimeEvent {
  event_id: string;
  run_id: string | null;
  step_id: string | null;
  event_type: RuntimeEventType;
  occurred_at: string;
  payload: Record<string, unknown>;
}

export type RuntimeEventType =
  | "command_received"
  | "route_selected"
  | "plan_created"
  | "run_created"
  | "agent_started"
  | "agent_completed"
  | "step_started"
  | "step_completed"
  | "step_failed"
  | "tool_call_started"
  | "tool_call_completed"
  | "tool_call_failed"
  | "approval_requested"
  | "approval_resolved"
  | "artifact_created"
  | "surface_created"
  | "fallback_triggered"
  | "run_completed"
  | "run_failed"
  | "run_cancelled";

export interface AgentWorkload {
  agent_name: string;
  call_count_24h: number;
  avg_duration_ms: number;
}

export interface RuntimeSummary {
  active_runs: number;
  blocked_runs: number;
  completed_24h: number;
  failed_24h: number;
  agents_active: number;
  top_agents: AgentWorkload[];
}

File: backend/src/api/schemas/runtime.py
"""Pydantic response models for runtime APIs."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class RuntimeStepResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")
    step_id: str
    status: str
    action: str | None = None
    tool_name: str | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    error: str | None = None


class RuntimeRunResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")
    run_id: str
    plan_id: str | None = None
    status: str
    created_at: datetime | None = None
    total_steps: int = 0
    completed_steps: int = 0
    progress_pct: int = 0
    blocking_step_id: str | None = None
    blocking_reason: str | None = None


class RuntimeEventResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")
    event_id: str
    run_id: str | None = None
    step_id: str | None = None
    event_type: str
    occurred_at: datetime
    payload: dict = Field(default_factory=dict)


class AgentWorkloadResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")
    agent_name: str
    call_count_24h: int
    avg_duration_ms: float


class RuntimeSummaryResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")
    active_runs: int = 0
    blocked_runs: int = 0
    completed_24h: int = 0
    failed_24h: int = 0
    agents_active: int = 0
    active_agents: list[str] = Field(default_factory=list)
    top_agents: list[AgentWorkloadResponse] = []
